### ISSUE TITLE
osnetwork.c API discrepancies

### ISSUE THREAD

**Issue by skliper** [nasa/osal#84]
The VxWorks osnetwork.c code differs from the OSAL Library API documentation under some conditions.  Even more interesting, the posix and rtems OSAL implementations behave differently from the VxWorks as well.

In osal/src/os/vxworks6/osnetwork.c, the documentation and code for

{{{
int32 OS_NetworkGetID (void) 
}}}
don't match when there is no network (e.g. OS_INCLUDE_NETWORK is undefined in osconfig.h).

The c source file function comments are:
{{{
  Returns: OS_ERROR if the  host id could not be found
    a 32 bit host id if success
}}}
The OSAL Library API.pdf has:
{{{
Returns:
OS_ERROR if the operating system calls fail
OS_SUCCESS if success  
}}}
''(That OS_SUCCESS looks like a copyNpaste error, all implementations return the host id for successful behavior.)''

But the VxWorks code returns a (valid) host id of zero, instead of OS_ERROR (a -1) or OS_ERR_NOT_IMPLEMENTED (osapi.h).

When looking at the osal/src/os/rtems/osnetwork.c, I see that it returns OS_ERR_NOT_IMPLEMENTED when the OS_INCLUDE_NETWORK is undefined in osconfig.h.  That return value isn't documented in the OSAL Library API.pdf or in any code comments.  The posix osnetwork.c pays no attention to OS_INCLUDE_NETWORK and always makes the OS-layer network calls (unlike the rtems and vxworks osal's).

Other API surprises: 
{{{
int32 OS_NetworkGetHostName (char *host_name, uint32 name_len)
}}}
can return OS_INVALID_POINTER with a NULL pointer arg.  Good implementation, but needs documentation.

Note: I didn't find a tie-breaker in the osapi-os-net.h because that common header had no comments.  Also, I was unable to find documentation (pdf or code comments) on the meaning of these osapi.h return codes.

'''Initial Recommendations for this ticket:'''
* All OSAL (osnetwork.c) implementations for OS_NetworkGetID() and OS_NetworkGetHostName() should return OS_ERR_NOT_IMPLEMENTED when the OS_INCLUDE_NETWORK is undefined and not even attempt to make OS calls.
* The OSAL Library API.pdf function documentation should be updated to cover all possible return values.
* The separate osnetwork.c documentation should be removed in favor of a common function description in the osapi-os-net.h, which all OSAL implementations use.  Its too easy to focus in on one .c source and not realize there is an API discrepancy.
* Document (in the osapi.h and in the OSAL Library API.pdf) the meanings of these common OSAL return codes.

**Comment by skliper** [comment]
_Imported from trac issue 61.  Created by abrown4 on 2015-06-26T18:28:38, last modified: 2016-01-26T12:48:23_

**Comment by skliper** [comment]
_Trac comment by abrown4 on 2015-07-09 14:11:08:_

Identified with...
commit: [changeset:5cc4968] Added coverage tests for osnetwork.c
commit: [changeset:9fccc17] Switching default osnetwork.c coverage build to 32-bit.

To reproduce issue #84 when testing, run "make NONET=TRUE" (see Makefile). 100% coverage.

**Comment by skliper** [comment]
_Trac comment by abrown4 on 2015-07-17 14:07:48:_

Current direction: return OS_ERR_NOT_IMPLEMENTED for the "no network" case for VxWorks OSAL only.

**Comment by skliper** [comment]
_Trac comment by abrown4 on 2015-07-17 18:02:49:_

commit: [changeset:82f215e] Fixed osnetwork return codes with no network.
Tested with coverage tests on Trac #45 branch and in CFS_TST 0d3a682.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-11-05 16:24:24:_

All changes above are cherry-picked and squash-committed to `development` branch

Commit [changeset:5572f09] available for CCB review, in new branch `trac-61-osnetwork-fixes`

**Comment by skliper** [comment]
_Trac comment by abrown4 on 2015-11-05 16:37:30:_

Joe, note on this ticket there are some recommended documentation changes and some recommended changes on the other OSALs.  (We didn't do those on #45 since we were focusing on the coverage testing and VxWorks updates only.)  Will those changes get handled on another ticket or this one?

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-11-05 16:50:13:_

Only focusing on code changes for the moment - the documentation topic is a good one to bring up the at the next CCB.  I suspect any ticket that also requires a doc update will need a separate action item prior to osal-4.2 release.

The change to other OSALs will also have to be decided if it is something to fix before osal-4.2 release or not.  I only saw actual code fixes for the vxwork6 implementation on most of these issues.

(Also keep in mind that the -ng variants for POSIX and RTEMS - trac #28 - already address almost everything here, but in a different way).

**Comment by skliper** [comment]
_Trac comment by sstrege on 2015-11-08 13:20:46:_

Approve/recommend the change set.

Is there a ticket for the recommended documentation changes and other OSAL changes?

**Comment by skliper** [comment]
_Trac comment by sduran on 2015-11-09 11:26:41:_

Approve.

There is not a separate ticket for the docs.  The need for the doc update that goes along with the code update was captured in this ticket.  If you want to add a separate ticket for the doc updates, be sure to reference this ticket.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-11-09 16:47:00:_

I recommend ACCEPT for the code changes.

I think the documentation updates should be part of this ticket i.e. we should not close this ticket until the documentation is also updated.  Since there is no "test" associated with documentation, this can stay in the `test_complete` state after we merge it (assuming we merge it) until the required doc updates are done.

Is anyone working on the documentation updates or is that an unassigned work item?

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-11-09 20:51:21:_

Recommend accept, with a little nervousness.

This one led me down some interesting and dusty pathways,
but it all came down to this: Host ID is a weird and less
than completely standardized thing. Zero is probably going
to appear, 0xFFFFFFFF is probably not going to appear, and
the arbitrary OS_ERR_NOT_IMPLEMENTED value (-28) happens to
be an unlikely value (but is remotely possible).

Someone would have to be really obnoxious to arrange for
their VxWorks implemtation to return FFFFFFE4. I think it
would require setting a network interface to think it was
assigned a multicast or subnet broadcast address (depending on
the endianness of the system).

But if we do this, please note, we are accepting a chunk
of technical debt: OS_ERR_NOT_IMPLEMENTED will have to
continue to NOT correspond to a sane valid Host ID for
the forseeable future.

It is unfortunate that the API did not split success and
failure indication away from returning the 32-bit opaque cookie.

**Comment by skliper** [comment]
_Trac comment by acudmore on 2015-11-10 10:04:19:_

I recommend accept. Will Joe's future updates to the error codes resolve the potential problem that Greg brought up?

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-11-10 11:07:49:_

Actually, we may have a correctness problem after all, just not in the VxWorks code.

If OS_NetworkGetID is intended to return an IPv4 address that might be used to contact the target in the future, the OSAL POSIX implementation is wrong, as "gethostid()" as used in the OSAL POSIX implementation of OS_NetworkGetID() does not return an IPv4 address on POSIX systems. We can get Linux to make gethostid() return our IPv4 address by writing it into /etc/hostid or calling sethostid(), but I see nothing along these lines in the code, and that would only cover Linux, not other POSIX targets.

http://www.unix.com/man-page/posix/3posix/gethostid/

http://linux.die.net/man/2/gethostid

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-11-16 13:07:44:_

Added comments to the header and implementations that I think boil down our discussions of last week to their essentials; see [changeset:ef8a740]. There was mention of actually including the GCC annotation that deprecates the call, but I don't recall that we decided to go forward on that. We would absolutely have to include sufficient cleverness so we did not break non-GCC (or ancient-GCC) builds ...

Inserted this comment:

{{{
OS_NetworkGetID is currently [[deprecated]] as its behavior is not 
always what could reasonably be expected: while it returns an IPv4  
address on VXWORKS (when OS_INCLUDE_NETWORK is defined), it does 
not do so on LINUX targets; for details, see the POSIX definition 
of gethostid() which OSAL uses to implement this function.
}}}

**Comment by skliper** [comment]
_Trac comment by sstrege on 2015-11-16 19:26:10:_

Overall this is good and gets the point across.  The comment addresses VxWorks and Linux but does not say anything in regard to other operating systems.  I think it would be worth changing slightly to be a bit more generic:

"OS_NetworkGetID is currently [[deprecated]] as its behavior is unknown and not consistent across operating systems.  The behavior of this API is not always what could reasonably be expected.  For example, while it returns an IPv4 ...." 

Also the OSAL Library API.doc and .pdf documents need to be updated with this comment

**Comment by skliper** [comment]
_Trac comment by abrown4 on 2015-11-17 09:59:55:_

BTW, the rtems version should be noted that it returns OS_ERR_NOT_IMPLEMENTED if OS_INCLUDE_NETWORK is not defined.  (These version-specific notes are quite useful in this current situation.)

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-11-20 13:10:23:_

Will trim back to just the first sentance of Susie's rewording and provide
a new commit id (Susie will update the API document).

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-11-20 16:12:41:_

now part of 2015-11-20

**Comment by skliper** [comment]
_Trac comment by sstrege on 2015-11-23 19:17:17:_

Updated API documents are attached and ready for review

**Comment by skliper** [comment]
_Trac comment by abrown4 on 2015-11-25 10:39:13:_

OSAL Library API.pdf review comments:
* There is a minor problem with the Network API formatting in the ToC, page 4.
* The OS_NetworkGetHostName also returns OS_ERR_NOT_IMPLEMENTED when OS_INCLUDE_NETWORK is not defined (VxWorks & RTEMS) - similar to OS_NetworkGetID.  Note, the POSIX implementation does not.

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-12-02 10:27:38:_

Just advanced development to include ic-2015-11-20 which includes the change to the source code, but which does not include the updated Library API document yet.

**Comment by skliper** [comment]
_Trac comment by glimes on 2016-01-14 12:34:59:_

Are we still waiting on an API document update?
Or did we just not get around to closing this ticket?

**Comment by skliper** [comment]
_Trac comment by glimes on 2016-01-19 13:09:13:_

Above API Doc file is present in [changeset:48d5d23],
but another version is coming soon.

**Comment by skliper** [comment]
_Trac comment by sstrege on 2016-01-20 14:13:01:_

Final revisions of API documents (.pdf and word) are attached and ready for merge to development

**Comment by skliper** [comment]
_Trac comment by glimes on 2016-01-21 10:24:41:_

Documents added to OSAL/doc in [changeset:1a547e1]
pushed as the `ic-trac-61-API-document` branch.

If I don't hear anything by tonight (and if I dohn't forget!)
I'll advance "development" to include this update.

**Comment by skliper** [comment]
_Trac comment by glimes on 2016-01-26 12:43:51:_

Why yes, I did get tied up in other things ...

Development on Babelfish and IPAS-LIB has been advanced
to include the [changeset:1a547e1] commit. Will finish
up this ticket and retire branches.
